### PR TITLE
【KernelGen】Add feature_dropout operator

### PR DESCRIPTION
**Pull request by Schopenhauer-loves-Hegel**
### PR Category
Operator

### Type of Change
New Feature

### Description
Add `feature_dropout` operator implementation with Triton kernel.

- Implementation mode: `manual_kernel`
- Accuracy test: 96/96 passed

### Issue
N/A

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [x] Change is fully covered by a UT.

### Performance
**torch.bfloat16**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| [8, 64] | 0.0128 | 0.0233 | 0.549 |
| [16, 128] | 0.0127 | 0.0223 | 0.571 |
| [32, 256] | 0.0128 | 0.0196 | 0.650 |
| [4, 64, 32, 32] | 0.0162 | 0.0214 | 0.757 |
| [8, 128, 16, 16] | 0.0162 | 0.0229 | 0.709 |
| [16, 256, 8, 8] | 0.0162 | 0.0216 | 0.750 |
| [2, 512, 14, 14] | 0.0161 | 0.0216 | 0.747 |
| [4, 1024, 7, 7] | 0.0190 | 0.0263 | 0.722 |

**torch.float16**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| [8, 64] | 0.0156 | 0.0389 | 0.401 |
| [16, 128] | 0.0156 | 0.0177 | 0.882 |
| [32, 256] | 0.0128 | 0.0231 | 0.553 |
| [4, 64, 32, 32] | 0.0180 | 0.0216 | 0.833 |
| [8, 128, 16, 16] | 0.0162 | 0.0220 | 0.739 |
| [16, 256, 8, 8] | 0.0163 | 0.0204 | 0.795 |
| [2, 512, 14, 14] | 0.0174 | 0.0219 | 0.798 |
| [4, 1024, 7, 7] | 0.0174 | 0.0229 | 0.761 |

**torch.float32**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| [8, 64] | 0.0140 | 0.0201 | 0.696 |
| [16, 128] | 0.0125 | 0.0264 | 0.474 |
| [32, 256] | 0.0140 | 0.0195 | 0.720 |
| [4, 64, 32, 32] | 0.0184 | 0.0272 | 0.676 |
| [8, 128, 16, 16] | 0.0155 | 0.0220 | 0.703 |
| [16, 256, 8, 8] | 0.0155 | 0.0203 | 0.762 |
| [2, 512, 14, 14] | 0.0164 | 0.0220 | 0.749 |
| [4, 1024, 7, 7] | 0.0151 | 0.0228 | 0.663 |

**Overall: median speedup = 0.721x, mean speedup = 0.694x** (24 data points)

---
_Generated by auto_gen tool with Claude Code_

### CODE REVIEW

**Comment by Schopenhauer-loves-Hegel** [comment]
Closing in favor of grouped PRs to reduce review overhead.

**Comment by factnn** [comment]
## feature_dropout Benchmark Speedup Results

### float16
| Size | Torch (ms) | Gems (ms) | Speedup |
|------|-----------|----------|---------|
| [8, 64] | 0.016 | 0.020 | 0.786 |
| [16, 128] | 0.016 | 0.018 | 0.898 |
| [4, 64, 32, 32] | 0.022 | 0.020 | 1.085 |
| [8, 128, 16, 16] | 0.017 | 0.020 | 0.835 |
| [2, 512, 14, 14] | 0.016 | 0.041 | 0.395 |
| [4, 1024, 7, 7] | 0.016 | 0.037 | 0.432 |

### float32
| Size | Torch (ms) | Gems (ms) | Speedup |
|------|-----------|----------|---------|
| [8, 64] | 0.014 | 0.037 | 0.384 |
| [4, 64, 32, 32] | 0.015 | 0.051 | 0.301 |
| [8, 128, 16, 16] | 0.016 | 0.044 | 0.375 |
| [4, 1024, 7, 7] | 0.017 | 0.048 | 0.344 |

### bfloat16
| Size | Torch (ms) | Gems (ms) | Speedup |
|------|-----------|----------|---------|
| [8, 64] | 0.014 | 0.045 | 0.304 |
| [4, 64, 32, 32] | 0.018 | 0.045 | 0.392 |
| [16, 256, 8, 8] | 0.016 | 0.034 | 0.482 |
| [4, 1024, 7, 7] | 0.016 | 0.042 | 0.382 |

Note: feature_dropout shows performance regression across most sizes, likely due to kernel launch overhead on small tensors.

File: benchmark/test_feature_dropout_perf.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    def set_shapes(self, shape_file_path=None):
+        self.shapes = [
+            (8, 64),
+            (16, 128),
+            (32, 256),
+            (4, 64, 32, 32),
+            (8, 128, 16, 16),
+            (16, 256, 8, 8),
+            (2, 512, 14, 14),
+            (4, 1024, 7, 7),
```
Comment (tengqm) [inline benchmark/test_feature_dropout_perf.py]: How did you choose these shapes?

File: src/flag_gems/__init__.py [hunk 2]
```diff
@@ -197,6 +197,8 @@ def torch_ge(v):
     ("expm1_", expm1_),
     ("expm1.out", expm1_out),
     ("exponential_", exponential_),
+    ("feature_dropout", feature_dropout),
+    ("feature_dropout_", feature_dropout_),
```
Comment (tengqm) [inline src/flag_gems/__init__.py:217]: Keep the list sorted, please

File: tests/test_feature_dropout.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    if QUICK_MODE
+    else [(2, 3), (4, 8, 16), (2, 16, 8, 8), (2, 32, 4, 4, 4)]
+)
+
+
+@pytest.mark.feature_dropout
+@pytest.mark.parametrize("shape", FEATURE_DROPOUT_SHAPES)
+@pytest.mark.parametrize("p", [0.3, 0.5, 0.7])
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
+def test_accuracy_feature_dropout(shape, p, dtype):
```
Comment (tengqm) [inline tests/test_feature_dropout.py]:
```suggestion
def test_feature_dropout(shape, p, dtype):
```

File: tests/test_feature_dropout.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    )
+    total = batch_size * num_channels
+    tolerance = max(0.3, 2.0 / (total**0.5)) if total < 50 else 0.2
+    assert abs(dropped / total - p) < tolerance
+
+
+@pytest.mark.feature_dropout
+@pytest.mark.parametrize("shape", FEATURE_DROPOUT_SHAPES)
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
+def test_accuracy_feature_dropout_no_train(shape, dtype):
```
Comment (tengqm) [inline tests/test_feature_dropout.py]:
```suggestion
def test_feature_dropout_no_train(shape, dtype):
```

File: tests/test_feature_dropout.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    with flag_gems.use_gems():
+        res_out = torch.feature_dropout(inp, 1.0, True)
+    gems_assert_equal(res_out, torch.zeros_like(inp))
+
+
+@pytest.mark.feature_dropout_
+@pytest.mark.parametrize("shape", FEATURE_DROPOUT_SHAPES)
+@pytest.mark.parametrize("p", [0.3, 0.5])
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
+def test_accuracy_feature_dropout_inplace(shape, p, dtype):
```
Comment (tengqm) [inline tests/test_feature_dropout.py]:
```suggestion
def test_feature_dropout_inplace(shape, p, dtype):
```

**Review by tengqm — COMMENTED** [review]
Overall lgtm.

File: benchmark/test_feature_dropout_perf.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        ]
+
+    def set_more_shapes(self):
+        return None
+
+
+@pytest.mark.feature_dropout
+def test_perf_feature_dropout():
```
Comment (tengqm) [inline benchmark/test_feature_dropout_perf.py]: Remove `_perf` from file name and function name

File: benchmark/test_feature_dropout.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+@pytest.mark.feature_dropout
+def test_perf_feature_dropout():
+    bench = FeatureDropoutBenchmark(
+        input_fn=feature_dropout_input_fn,
+        op_name="feature_dropout",
+        torch_op=torch.feature_dropout,
+        dtypes=FLOAT_DTYPES,
+    )
+    bench.run()
```
Comment (tengqm) [inline benchmark/test_feature_dropout.py:20]: Missing benchmark for inplace version

File: src/flag_gems/ops/feature_dropout.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+    with torch_device_fn.device(device):
+        apply_feature_mask_kernel[grid_apply](
+            input, out, mask, numel, N, C, spatial_size, BLOCK
+        )
+
+    return out
+
+
+def feature_dropout_(input, p, train=True):
```
Comment (tengqm) [inline src/flag_gems/ops/feature_dropout.py:185]: I believe the in-place version can be drastically simplified.

File: PR_status.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
+# Feb 27 KernelGen PR Status
```
Comment (tengqm) [inline PR_status.md]: Don't submit this markdown file.

**Review by tengqm — DISMISSED** [review]

**Review by tengqm — COMMENTED** [review]
lgtm

File: benchmark/test_feature_dropout.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
+import pytest
+import torch
+
+from benchmark.attri_util import FLOAT_DTYPES
+from benchmark.performance_utils import GenericBenchmarkExcluse1D
```
Comment (tengqm) [inline benchmark/test_feature_dropout.py]: please rebase this to master.
The `attri_util` is gone, the `performance_utils` is gone.
See `https://github.com/flagos-ai/FlagGems/blob/master/benchmark/test_addcdiv.py` for an example for the latest test case code sample.

**Review by tengqm — DISMISSED** [review]

**Comment by factnn** [comment]
## UT Results

Branch rebased to upstream/master and tests passed:

```
python -m pytest tests/ -m feature_dropout --ref cpu -x -q
72 passed
```

**Review by bin913 — APPROVED** [review]
